File: events/2016/05/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";
import * as CryptoJS from "crypto-js";

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): string | undefined {
  let doorID = input[0]
  let salt = 0;
  let password: string = "";

  while (password.length <= 7) {
    let hashIn = doorID + salt
    
    let hash = CryptoJS.MD5(hashIn).toString();
    if (hash.slice(0, 5) === "00000") {
      password += hash[5]
    }
    salt += 1
  }

  return password;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): string | undefined {
  let doorID = input[0]
  let salt = 0
  let password = ""
  let password_parts: Record<number, string> = {}

  for (var i = 0; i < 8; i++){
    password_parts[i] = ""
  }

  while (password.length <= 7) {
    let hashIn = doorID + salt
    
    let hash = CryptoJS.MD5(hashIn).toString();
    if (hash.slice(0, 5) === "00000") {
      let location = parseInt(hash[5])
      if (password_parts[location] == "") {
        password_parts[location] = hash[6]
      }
    }
    salt += 1

    password = ""
    for (var i = 0; i < 8; i++){
      password += password_parts[i]
    }
  }
  return password;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2016,
  Day: 5,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));

File: events/2016/05/solution.go
package main

import (
  "encoding/json"
  "fmt"
  "os"
  "time"
  "yod_aoc/utils/go/aoc_utils"
)

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
func part1(input []string) int {
  return -1
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
func part2(input []string) int {
  return -1
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
func main() {
  // Read the arguments
  args := os.Args[1:]
  var fileName string
  if len(args) > 0 {
    fileName = args[0]
  } else {
    fileName = "input.txt"
  }
  // Start the timer!
  startTime := time.Now()

  // Read the input to an array of strings
  path := aoc_utils.GetFilePath(2016, 05, fileName)
  content, err := aoc_utils.ReadInputAsArray(path)
  if err != nil {
    fmt.Println("Error reading input:", err)
    return
  }

  // Run the Parts
  p1Result := aoc_utils.MeasurePerformance(func() interface{} {
    return part1(content)
  })

  p2Result := aoc_utils.MeasurePerformance(func() interface{} {
    return part2(content)
  })

  // End the Timer!
  endTime := time.Now()
  duration := endTime.Sub(startTime)

  resultsForDay := aoc_utils.AOCDayResults{
    Year:     2016,
    Day:      05,
    Part1:    p1Result,
    Part2:    p2Result,
    Duration: duration.Nanoseconds(),
    Timestamp: aoc_utils.AOCTimestamp{
      Start: startTime.Format(time.RFC3339),
      End:   endTime.Format(time.RFC3339),
    },
  }
  jsonResults, _ := json.Marshal(resultsForDay)
  fmt.Println(string(jsonResults))
}
